File: median_sorted_arrays.py
# -*- coding: utf-8 -*-
"""
TC: O(log(m+n)) whre m = length of nums1 and n = length of nums2
SC: O(1) no extra space used
"""

class Solution:
    def findMedianSortedArrays(self, nums1:[], nums2:[]):
        #edge case
       
        n1 = len(nums1)
        n2 = len(nums2)
        total = (n1 + n2)
        half = total //2
        
        if n1 > n2:
            nums1,nums2 = nums2,nums1
        
        #Assign left and right pointer
        l,r = 0, len(nums1)-1
        
        while True:
            i = (l + r) //2
            j = half - i - 2
            
            Aleft = nums1[i] if i >= 0 else float("-infinity")
            Aright = nums1[i+1] if (i+1) < len(nums1) else float("infinity")
            Bleft = nums2[j] if j >= 0 else float("-infinity")
            Bright = nums2[j+1] if (j+1) < len(nums2) else float("infinity")
            
            #partition is correct
            if Aleft <= Bright and Bleft <= Aright:
                
                if total % 2:
                    return min(Aright,Bright)
                else:
                    return (max(Aleft,Bleft) + min(Aright,Bright))/2
                    
            elif Aleft > Bright:
                r = i - 1
            else:
                l = i + 1     
                

        

File: medianSortedArrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        
        int n1 = nums1.length;int n2 = nums2.length;
         if(n1 > n2){
             return findMedianSortedArrays(nums2,nums1);
         }
        int low = 0;int high = n1;
        
        while(low <= high){
            int partx = low + (high - low)/2;
            int party= (n1+n2)/2 - partx;
            double l1 = partx == 0 ? Integer.MIN_VALUE : nums1[partx - 1];
            double r1 = partx == n1 ? Integer.MAX_VALUE : nums1[partx];
            double l2 = party == 0 ? Integer.MIN_VALUE : nums2[party-1];
            double r2 = party == n2 ? Integer.MAX_VALUE : nums2[party];
            
            if(l1 <= r2 && l2 <= r1){
                if((n1+n2)%2 == 0 )
                    return (Math.max(l1,l2) + Math.min(r1,r2)) / 2;
                else
                    return Math.min(r1,r2);
            }
            if(l2 > r1){
                low = partx + 1;
            }
            if(l1 > r2){
                high = partx - 1;
            }
        }
        return 0.0;
    }
}
